Allow pip install with entrypoint names with colons

montefra opened the pull request:
I did try to install a package (from a local pypi-like server) containing colons in entry point names and the installation crashed after copying all the files to the site-packages directory but before creating the entry points:

```
Exception:
Traceback (most recent call last):
  File "lib/python3.4/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "lib/python3.4/site-packages/pip/commands/install.py", line 317, in run
    prefix=options.prefix_path,
  File "lib/python3.4/site-packages/pip/req/req_set.py", line 742, in install
    **kwargs
  File "lib/python3.4/site-packages/pip/req/req_install.py", line 831, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "lib/python3.4/site-packages/pip/req/req_install.py", line 1032, in move_wheel_files
    isolated=self.isolated,
  File "lib/python3.4/site-packages/pip/wheel.py", line 352, in move_wheel_files
    console, gui = get_entrypoints(ep_file)
  File "lib/python3.4/site-packages/pip/wheel.py", line 229, in get_entrypoints
    cp.readfp(data)
  File "/usr/lib/python3.4/configparser.py", line 735, in readfp
    self.read_file(fp, source=filename)
  File "/usr/lib/python3.4/configparser.py", line 690, in read_file
    self._read(f, source)
  File "/usr/lib/python3.4/configparser.py", line 1069, in _read
    fpname, lineno)
configparser.DuplicateOptionError: While reading from '<???>' [line 10]: option 'common' in section 'section' already exists
```

the entry points look like

```
[section]
common:one = module:func
common:two = module:other_func
```

This is a similar issue to #3434.
If you accept this change I'll add tests and change logs.

As a temporary work around, is there are way to avoid going through `pip/wheel.py`?

Also if I run `pip install -e .` in the source code directory it works fine (because it doesn't use the wheel.py module).

xavfernandez commented:
Th change seems sensible.
To avoid using wheel you can use the `--no-binary=yourpackage` option if a `.tar.gz` file is available.

montefra commented:
Great. I'll keep working on this in the next few days.

The ``--no-binary=yourpackage`` works fine, thank you.

When do you plan the next release?

montefra commented:
Test added.
With the changes to the test_requires in the ``setup.py`` file I managed to successful run ``tests/unit/test_wheel.py`` with the changes.

If I try to run ``tox`` I get:
```
Could not find platform dependent libraries <exec_prefix>                                                                                                                                            
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]                                                                                                                                             
Traceback (most recent call last):                                                                                                                                                                   
  File "/data/Codes/Gitrepos/pip/.tox/docs/bin/pip", line 7, in <module>                                                                                                                             
    from pip import main                                                                                                                                                                             
  File "/data/Codes/Gitrepos/pip/.tox/docs/local/lib/python2.7/site-packages/pip/__init__.py", line 13, in <module>
    from pip.utils import get_installed_distributions, get_prog
  File "/data/Codes/Gitrepos/pip/.tox/docs/local/lib/python2.7/site-packages/pip/utils/__init__.py", line 21, in <module>
    from pip.compat import console_to_str, expanduser, stdlib_pkgs
  File "/data/Codes/Gitrepos/pip/.tox/docs/local/lib/python2.7/site-packages/pip/compat/__init__.py", line 27, in <module>
    import sysconfig
  File "/usr/lib64/python2.7/sysconfig.py", line 10, in <module>
    'stdlib': '{base}/'+sys.lib+'/python{py_version_short}',
AttributeError: 'module' object has no attribute 'lib'
```

I've tried to run pytest in a py35 virtual environment but I get a few failures, some of which are related with mercurial and bzr, that I don't have on the machine I'm currently using.

``py.test --lf -x -m 'not mercurial and not bzr'``

correctly disable some test but not all the tests using them are correctly marked.

montefra commented:
I'm fixing the errors in python2.x.

Also tox works fine on a different machine.

montefra commented:
The problem is bigger than I though.

Python 2 ``ConfigParser`` use ``=`` and ``:`` without any option to chose between the two. The only way around it is to inherit from ``ConfigParser.RawConfigParser`` and override the [``OPTCRE*``](https://hg.python.org/cpython/file/2.7/Lib/ConfigParser.py#l446) class variables.
It just doesn't fail because python2 accepts duplicate options, while python3 chokes on them.

``OPTCRE*`` are public constant variables and look the same in Python 2.6 and 2.7.

This is the second time that I fix a this kind of issue and I wouldn't be surprised if this will come up again in the future.

I deem that we have two options:

* Add a module that exposes the three ``*ConfigParser`` classes, for all the supported versions of python, but with ``=`` separator enabled.
* parse the ``entry_points.txt`` file in the same way as setuptools does. This would apply to ``wheel.py`` and in the [uninstall command](https://github.com/pypa/pip/blob/master/pip/req/req_install.py#L716). I can try to figure out how this work, unless you already know.

I would be hesitant in banning entry points with colons in the name since setuptools accepts and correctly parse them.

xavfernandez commented:
@montefra you could maybe use https://github.com/pypa/pip/blob/84c969669c0242c9e70137312b7d1001b4af26f8/pip/_vendor/pkg_resources/__init__.py#L2300-L2316

montefra commented:
@xavfernandez : thank you for the pointer. I'll look into it.

montefra commented on pip/wheel.py:
```diff
@@ -225,20 +225,19 @@ def get_entrypoints(filename):
+
+    # convert the EntryPoint objects into strings with module:function
+    console = dict(_split_ep(v) for v in console.values())
+    gui = dict(_split_ep(v) for v in gui.values())
```
@xavfernandez : if you approve this implementation, I'll use the entry point mechanism also in the [uninstall](https://github.com/pypa/pip/blob/master/pip/req/req_install.py#L716) implementation.

montefra commented:
I have added the use of the setuptools entry points when uninstalling a package and updated the corresponding test. Tox works fine except for the case of py2.7 where I get:

    [...]
    tests/functional/test_install.py .F..................Xs.....................................ss..
    [...]
```
    def test_with_setuptools_and_import_error(script, data):
        # Make sure we get an ImportError while importing setuptools
        setuptools_init_path = script.site_packages_path.join(
            "setuptools", "__init__.py")
        with open(setuptools_init_path, 'a') as f:
            f.write('\nraise ImportError("toto")')
    
        result = script.run(
            "python", "-c",
            "import pip; pip.main(["
            "'install', "
            "'INITools==0.2', "
            "'-f', '%s', "
            "'--no-binary=:all:'])" % data.packages,
            expect_error=True,
        )
>       assert (
            "Could not import setuptools which is required to install from a "
            "source distribution."
            in result.stderr
        )
E       assert 'Could not import setuptools which is required to install from a source distribution.' in ''
E        +  where '' = <tests.lib.TestPipResult object at 0x7f48fbf6aad0>.stderr
```

The failure goes away if I run ``tox -e py27 -- tests/functional/test_install.py``. The ``X`` test is:

    tests/functional/test_install.py::test_install_argparse_shadowed XPASS

I was trying to figure out if this happen also on travis, but I have not found if ``test/functional`` is run at all.

montefra commented:
``tests/functional/test_install.py::test_install_argparse_shadowed`` ``XPASS`` also in commit a88beb1523f5c5ac9703b7c13296058d3725d54a (so it's not my fault).

I am investigating which commit made the other test fail.

montefra commented on pip/req/req_install.py:
```diff
@@ -715,32 +714,18 @@ def uninstall(self, auto_confirm=False):
-        if dist.has_metadata('entry_points.txt'):
-            if six.PY2:
-                options = {}
+        console_scripts = dist.get_entry_map(group='console_scripts')
```
@xavfernandez shouldn't this also be done for ``gui_scripts``?

xavfernandez replied:
Good point, I'm wondering if the gui_scripts ar correctly uninstalled...
But I guess they should be in the `installed-files.txt` file.

montefra replied:
Do you want a new PR for this?

xavfernandez replied:
It might be simpler to add the `gui_scripts` case in this PR.

montefra replied:
Ok, I'll do it either tomorrow or after mid next week.

montefra commented:
@xavfernandez : should I merge master into this branch?

xavfernandez commented on setup.py:
```diff
@@ -42,7 +42,7 @@ def find_version(*file_paths):
 
 tests_require = ['pytest', 'virtualenv>=1.10', 'scripttest>=1.3', 'mock',
-                 'pretend']
+                 'pretend', 'pytest-catchlog', 'freezegun']
```
This should not be in this PR

montefra replied:
I did for some reason I don't recall (possibly because I had to install something by hand to make pytest work). I can revert it. Do you want a new PR for this?

xavfernandez commented:
@montefra Sorry for the delay, I was in holidas.
I'd prefer that you rebase on master, which will avoid needless merge commits.

montefra commented:
@xavfernandez : I'll do the rebase in the next few days

montefra commented:
@xavfernandez : I've rebased the branch on master and reverted the changes on setup.py.

Do you want me to add ``console_scritps`` into ``pip/req/req_install.py:717`` in this or an other PR?

montefra commented:
@xavfernandez : I have added gui_scripts to the unistall part and I haven't broken any test.

However it is not clear to me how I could test that gui_scripts get installed and removed correctly.

Any pointer on what file or package in ``test/data`` I should modify?

xavfernandez commented on tests/functional/test_uninstall.py:
```diff
@@ -166,16 +169,20 @@ def test_uninstall_entry_point(script):
+    """.format(console_scripts)))
+    script_name = script.bin_path.join(console_scripts.split('=')[0].strip())
     result = script.pip('install', pkg_path)
+    assert script_name.exists
```
is `exist` a property or a method ?

montefra replied:
It's a property:

    (Pdb) p type(script_name)
    <class 'tests.lib.path.Path'>
    (Pdb) p script_name.exists
    True

xavfernandez commented:
@montefra To test this, I'd go with a custom package whose setup.py you define in the test, just like for `test_uninstall_entry_point` test.

montefra commented:
@xavfernandez : I'll do it soon.

xavfernandez commented on tests/functional/test_uninstall.py:
```diff
@@ -166,20 +169,47 @@ def test_uninstall_entry_point(script):
+            entry_points={{"gui_scripts": ["test_ = distutils_install", ],
+                          }}
+        )
+    """.format(pkg_name)))
```
With a recent rebase on master, this could be rewritten with `from tests.lib import create_test_package_with_setup
`, cf https://github.com/pypa/pip/blob/a12f2f493cc705f3acae104cf4dda5d0db519cf8/tests/functional/test_check.py#L25-L28

montefra replied:
I'm working on it.
However not hat I have rebased for some reason I cannot run ``tox`` on my machine. I get errors like

> tox/py35/bin/python: Error while finding spec for 'pip.__main__' (<class 'ImportError'>: bad magic number in 'pip._vendor.html5lib.serializer': b'\x03\xf3\r\n'); 'pip' is 
a package and cannot be directly executed

I'll see what the CIs tell me and act accordingly

montefra commented:
I've rebased on top of master and pushed to my repo.
However I forgot to push with ``--force`` and something has gone weird with the history of the branch.
Any idea on how to fix this?

I'll see if I manage to update the test as per your comment.

montefra commented:
I did:

* git checkout master
* git fetch upstream
* git merge upstream/master
* git push origin master
* git checkout fix_install_colon
* git rebase master
and got:
```
First, rewinding head to replay your work on top of it...
Applying: use only = delimiter when creating a wheel
Applying: add pytest-catchlog and freezegun
Applying: test entry points with colon in name
Applying: Add item to change log file
Using index info to reconstruct a base tree...
M       CHANGES.txt
Falling back to patching base and 3-way merge...
Auto-merging CHANGES.txt
Applying: fix format syntax in python2.6
Applying: load the entry points using pkg_resources
Applying: Fix PEP8
Applying: uninstall: use pkg_resource distribution when getting the entry point names
Applying: revert commit 7c45aeb546e35f3b77c3fb04c3bb6d78036256cf as per request
Applying: add console_scripts names to the list to uninstall
Applying: Add test to uninstall GUI scripts
```
I have fixed the conflict and committed

Then pushed and the mess happened.

Now I have tried an other rebase and I get a different error. I'll report later

montefra commented:
I run ``git pull`` instead of ``git push`` by error and it seems that that caused the mess. [This stackoverflow answer](http://stackoverflow.com/a/30927009/1860757) seems very good.

I see what I can do. Probably the safest way is to create a new branch on top of current trunk, cherrypick all my changes and create a new PR.

xavfernandez commented:
> I have fixed the conflict and committed
>
> Then pushed and the mess happened.

When you get a conflict on a rebase, you are supposed to fix the conflict, `git add` the resolved files and `git rebase --continue`, not `git commit`. So this might be the reason...

You could use `git reflog` to go back to before this issue.

xavfernandez commented:
If https://github.com/pypa/pip/commit/5f20204 was your last good commit, you can:

    git checkout fix_install_colon
    git reset --hard 5f2020414008b0aa570c424e33d2f8a3c12ef5cb
    git push -f

To go back to this previous state.

montefra commented:
> When you get a conflict on a rebase, you are supposed to fix the conflict, git add the resolved files and git rebase --continue, not git commit

I know :(. But sometimes you do something stupid, isn't it? Also the fact that I switch between git and svn on regular base doesn't help :(

> You could use git reflog to go back to before this issue.

I can also try this.

montefra commented:
> git reset --hard 5f2020414008b0aa570c424e33d2f8a3c12ef5cb
> git push -f

it did work, thank you very much. Now I'll try to avoid an other mess :smiley:

montefra commented:
Correctly rebased. :sweat_smile: 
Now I'll update the changelog and the tests using the function you showed me before.

xavfernandez commented:
@montefra Thanks a lot for your work :+1:

montefra commented:
Thank you for accepting the PR and helping with the git-mess :+1:
I hope that I'm done with colons in ``pip`` :wink:
